echo_server.py

```diff
@@ -1,28 +1,149 @@
 #! /usr/bin/env python
 """Echo server in socket connection: receives and sends back a message."""
 import socket
+import os
+import mimetypes
+import binascii
+
+
+def response_ok(tuple):
+    """Return byte string 200 ok response."""
+    # u"HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-length: 18\r\n\r\neverything is okay".encode('utf-8')
+    okay = u"HTTP/1.1 200 OK\r\nContent-Type: "
+    c_len = u"\r\nContent-length: "
+    crlf = u"\r\n\r\n"
+    return "{}{}{}{}{}{}".format(okay, tuple[0], c_len, len(tuple[1]), crlf, tuple[1])
+
+def response_error(error_code, reason):
+    """Return byte string error code."""
+    return u"HTTP/1.1 {} {}\r\nContent-Type: text/plain\r\nContent-length: {}\r\n\r\n{}".format(error_code, reason, len(reason), reason).encode('utf-8')
+
+
+# def parse_request(request):
+#     """Parse HTTP request and return URI requested.
+#     If not GET requests, raises error.
+#     If not HTTP/1.1 request, raises error."""
+#     first_line = request.split("\n")[0].split()
+#     if first_line[2] == 'HTTP/1.1':
+#         if first_line[0] == 'GET':
+#             return first_line[1]
+#         else:
+#             return response_error(405, 'Method Not Allowed')
+#     else:
+#         return response_error(505, 'HTTP Version Not Supported')
+
+class Error404(BaseException):
```

Comment:
It's traditional to inherit from `Exception` rather than `BaseException`.  I believe there's actually a reasonably good reason for it, but cannot remember what it is.